src/servers/src/http/event.rs

```diff
@@ -0,0 +1,241 @@
+pub struct LogIngesterQueryParams {
+    pub table: Option<String>,
+    pub db: Option<String>,
+    pub pipeline_name: Option<String>,
+    pub ignore_errors: Option<bool>,
+
+    pub version: Option<String>,
+}
+
+pub struct PipelineContent(String);
+
+#[async_trait]
+impl<S, B> FromRequest<S, B> for PipelineContent
+where
+    B: HttpBody + Send + 'static,
+    B::Data: Send,
+    bytes::Bytes: std::convert::From<<B as HttpBody>::Data>,
+    B::Error: Into<BoxError>,
+    S: Send + Sync,
+{
+    type Rejection = Response;
+
+    async fn from_request(req: Request<B>, state: &S) -> StdResult<Self, Self::Rejection> {
+        let content_type_header = req.headers().get(CONTENT_TYPE);
+        let content_type = content_type_header.and_then(|value| value.to_str().ok());
+        if let Some(content_type) = content_type {
+            if content_type.ends_with("yaml") {
+                let payload = String::from_request(req, state)
+                    .await
+                    .map_err(IntoResponse::into_response)?;
+                return Ok(Self(payload));
+            }
+
+            if content_type.starts_with("multipart/form-data") {
+                let mut payload: Multipart = Multipart::from_request(req, state)
+                    .await
+                    .map_err(IntoResponse::into_response)?;
+                let file = payload
+                    .next_field()
+                    .await
+                    .map_err(IntoResponse::into_response)?;
+                let payload = file
+                    .ok_or(StatusCode::UNSUPPORTED_MEDIA_TYPE.into_response())?
+                    .text()
+                    .await
+                    .map_err(IntoResponse::into_response)?;
+                return Ok(Self(payload));
+            }
+        }
+
+        Err(StatusCode::UNSUPPORTED_MEDIA_TYPE.into_response())
+    }
+}
+
+#[axum_macros::debug_handler]
+pub async fn add_pipeline(
+    State(handler): State<LogHandlerRef>,
+    Path(pipeline_name): Path<String>,
+    Extension(query_ctx): Extension<QueryContextRef>,
+    PipelineContent(payload): PipelineContent,
+) -> Result<String> {
+    if pipeline_name.is_empty() {
+        return Err(InvalidParameterSnafu {
+            reason: "pipeline_name is required in path",
+        }
+        .build());
+    }
+
+    if payload.is_empty() {
+        return Err(InvalidParameterSnafu {
+            reason: "pipeline is required in body",
+        }
+        .build());
+    }
+
+    let content_type = "yaml";
+    let result = handler
+        .insert_pipeline(&pipeline_name, content_type, &payload, query_ctx)
+        .await;
+
+    result.map(|_| "ok".to_string()).map_err(|e| {
+        error!(e; "failed to insert pipeline");
+        e
+    })
+}
+
+/// Transform NDJSON array into a single array
+fn transform_ndjson_array_factory(
+    ignore_error: bool,
+) -> impl FnMut(Result<Value>, std::result::Result<Value, serde_json::Error>) -> Result<Value> {
```

Comment:
We could refactor the function to take an iterator of `Value`. It should be more clear than a function returning a closure to be used in the `fold()` method.
```suggestion
fn transform_ndjson_array_factory(
    values: impl Iterator<Item = Value>,
    ignore_error: bool,
) -> Result<Value> {
```